text/0095-add-latest-cf.md

```diff
@@ -0,0 +1,72 @@
+- key set the original key without any encoding or version
+- value set the same value in write cf but include the corresponding version.
+
+For example, insert k1 with version v0 and value dummy will insert two keys
+- to write cf, k1|v0 -> (dummy and other meta)
+- to latest cf, k1 -> (dummy, v0 and other meta)
+
+So all keys in latest cf represent the latest version of all keys.
+
+When all the versions of a key are gced, then it should also delete the key in latest cf only when it matches the last gc key version.
+
+When a key is queried, it should query latest cf using `get` first. If nothing is found or the version is larger than requested, it should query the write cf as fallback. In most case, only one `get` is performed.
+
+When a range scan is triggered, it should scan the latest cf directly. If a larger version is met, it should fallback to seek write cf for that specific key instead. Because only the latest versions are stored in latest cf, so the keys needs to be scanned will be way fewer than the write cf. And in most cases, only one `seek` is performed, and all other operations are `next`.
+
+All of the fallback queries should be performed lazily.
+
+The improvment should be very significant when update keys frequently.
+
+### Compatibility
+
+Because all existing cfs are updated just as before, so there are no major compatibility issues.
+
+But using latest cf should be triggered explicitly by client. Client should ensure only when it updates all keys with latest cf will it ask TiKV to query using latest cf.
+
+Take TiDB as an example, it can add a new storage format at table level. Perhaps even add a new DDL job for table to changing the storage format. In the new format, latest cf is always updated. And only trigger TiKV to use latest cf when the target table is fully upgraded to the new format.
+
+### Why use a new cf?
+
+If the latest keys are written to write cf instead, then it will break compatibility. It also makes range scan less efficient as more version need to be scanned and skipped.
+
+## Drawbacks
+
+It introduces a write amplification appearantly. That is also the reason why the access pattern needs to be controlled by client. Client should enable latest cf only when it knows a range of keys are updated very often and can be benificial from the change.
+
+On the other hand, the additional write is just a key in a different cf and a value that is probably not larger than 255 bytes, the overhead may not be very signifiant. More experiments are needed.
+
+## Alternatives
+
+unistore separates the latest version and other versions by adjust file format. So when flushing or compacting, it will make latest versions key be the first part, and the rest in the second part. This approach doesn't have write overhead, but is not backward compatible in TiKV.
```

Comment:
How about create a history cf to store as many history version as it can, the write cf store the latest several versions, and background thread like GC thread move existed history versions to history cf.
For point get, using RocksDB's user timestamp feature to get the latest version can see, this also can prevent creating iterator.

Reply:
There is no compatibility issue for this solution.

Reply:
Compaction filter is not reliable, we have seen issue cause by not having compaction in time like https://github.com/tikv/tikv/issues/12729. On the other hand, this proposal doesn't depend on the user timestamp, which is not used in production yet.

Reply:
Added as an alternative.

Reply:
Using history cf doesn't need the compaction filter, gc just check the history cf sst file's max ts to determine wether to drop the whole sst file.

Reply:
Compaction filter is not used for dropping history sst, but for moving data from write cf to history cf.

Reply:
Compaction filter is not necessary, I think version statistics-based GC is much better.